### ISSUE TITLE
Why decode?

### ISSUE THREAD

**Issue by rhpvorderman** [marcelm/dnaio#11]
Hi Marcel,

I got a bit fed up with the crappy fastq filters in the bioinformatics space so I started on something that is fast and actually evaluates quality scores correctly (taking into account they are log values, so many people simply compute the arithmetic mean...). 

Currently I implemented a very steady and simple pure Python parser that constructs a NamedTuple containing the 4 lines (without `\n`) as bytes objects. It is about 2.5 times slower as dnaio.

Bytes objects are nice, since quality scores are integers and bytes objects can also be represented in python as a sequence of integers directly.  (You can sum bytes, throw them into statistics.median etc. with no issue).

The same goes for the sequence string. AGTCN etc. are all in the ASCII range, so there is no reason to represent it in unicode. 

As for cutadapt, bytes have a buffer protocol that allow for some very mean and lean reading of their internal uint8_t array. Strings do not have this. This can be quite powerful see [this function for mean quality](https://github.com/LUMC/fastq-filter/blob/49d485fcb8a3e7f49739df2e53a661ee15cb3483/src/fastq_filter/_optimize.pyx#L10).

So I wonder why decode is used at all? Wouldn't it be more convenient for Sequence objects to have members as bytes instead? No decoding -> faster parsing and easier usage with other libraries that use cython (i.e. cutadapt).

Maybe it would be possible to set up a BinarySequence object which is parsed faster. For backwards compatibility this could be converted into Sequence object by default. What do you think on this issue?

**Comment by marcelm** [comment]
When I converted Cutadapt from Python 2 to 3, I thought about this for quite a long time. In Python 2, strings were of course ASCII-only, and because changing them to unicode would make operations on them slower, I actually used bytes objects in the beginning. Then in commit https://github.com/marcelm/cutadapt/commit/6df01765880b5ec983bf4c85b6f681d1155125f2, I switched to regular str objects. The commit message doesn’t say this, but the main reason was that the code that did anything with the sequences became harder to read. This may have also had to do with trying to maintain Python 2 and 3 compatibility in the same code base, but some things remain, in particular for the nucleotide sequence itself:
- You need to remember to write single nucleotides as b'A'
- When you index, `seq[i]` gives you an int, so you need to remember to use `seq[i:i+1]` every time.
- `print(b'ACGT')` prints out `b'ACGT'`, so you need to remember to convert to str every time you print something to the user.

For myself, I came to the conclusion that nucleotide sequences *are* indeed strings or that at least I want to be able to use them like strings.

Regarding performance: I have recently benchmarked this once again and `.decode()` and `.encode()` are *extremely* fast for ASCII-only strings, at least in recent Python versions. As you may have seen in the commit message, I refer to PEP 393 (Flexible String Representation). I don’t know if you are aware, but latin-1 (and consequently, ASCII-only) strings are stored internally the same way a Python 2 string was stored (one byte per character), so the only thing `.decode()` needs to do is to inspect each character. And by the way, if you use `.decode("latin-1")`, it’s even faster than when you let it use the default encoding UTF-8.

Regarding the quality values, the story is a bit different because you actually want to do computations with them as you said. So I agree it would be quite useful to store them in such a way that indexing one element gives you an int. I don’t know whether a bytes object would be the best option or possibly an array.

However, I also think that indexing into the qualities should give you the actual phred-scaled quality and not quality plus 33. Back when pysam did an API overhaul, I suggested this to the developers and this is now [how it’s implemented](https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.query_qualities).

Ah, and I just found this discussion where I already mention some of the thoughts above: https://github.com/pysam-developers/pysam/issues/29.

So in short: I’m open to changing `.qualities` somehow, but not quite for `.sequence`.

**Comment by rhpvorderman** [comment]
Thank you for your comprehensive answer. I learned a lot from it.

Still, there is no need to decode. It is possible to write the ascii strings directly in python objects. Since ASCII is exactly the same as UTF-8. So I [tested that](https://github.com/rhpvorderman/dnaio/commit/8ee86ea48775b2973bda89d1cc949445bf4584e0). It works, but the performance gains are zero (I couldn't measure a difference). So it is not worth losing the checking that the decode provides.

It turns out there is no way to further improve the speed of dnaio. I have been looking but it is a "where is Waldo" quality puzzle and I haven't found waldo yet :wink:.

> Regarding the quality values, the story is a bit different because you actually want to do computations with them as you said. So I agree it would be quite useful to store them in such a way that indexing one element gives you an int. I don’t know whether a bytes object would be the best option or possibly an array.

> However, I also think that indexing into the qualities should give you the actual phred-scaled quality and not quality plus 33. Back when pysam did an API overhaul, I suggested this to the developers and this is now how it’s implemented.

I want to note that calculations are much faster if you subtract the phred offset from the resulting value afterwards instead of subtracting it from each element beforehand. Simple example: median. `statistics.median(quality_bytes) - phred offset` returns the correct answer and is much faster than `statistics.median(qual - phred_offset for qual in quality_bytes)`. So it is more performant to have all functions that do something with quality have a `phred_offset` variable instead of handling it in the parser.

Anyway. Keeping qualities as strings is also okay, since encode('latin-1') is blazing fast anyway, there is no real need to keep them as bytes. Keeping them as strings has advantages during printing and writing to new files.

Thanks again for all the background information!

**Comment by marcelm** [comment]
> Still, there is no need to decode. It is possible to write the ascii strings directly in python objects. Since ASCII is exactly the same as UTF-8. So I tested that. It works, but the performance gains are zero (I couldn't measure a difference). So it is not worth losing the checking that the decode provides.

Cool, thanks for testing that! I had actually thought about doing something like this, but then I think I never found the correct API call to use.

> I want to note that calculations are much faster if you subtract the phred offset from the resulting value afterwards instead of subtracting it from each element beforehand. 

Yes, that’s a good point. I think that in principle I still would want to do the conversion when reading in the file (or actually it could be an attribute that is lazily computed on first access) because I would want to "shield" the user from the implementation details. One more thought regarding that: It would make sense to support reading of BAM files in dnaio as well, and these store qualities without offset.

By the way, on the topic of quick parsing, I can recommend this talk if you have the time: https://www.youtube.com/watch?v=wlvKAT7SZIQ (I wanted to write a FASTQ in assembly after I had seen it, but that would take me too much time.)

Closing this now, but feel free to open PRs with speed improvements :smile:.

**Comment by rhpvorderman** [comment]
Ah thanks again for the information and the pointer to the talk. If I find some speed improvement in dnaio I will surely suggest it. (But you have made this really, really, really hard! :wink:).
